server/master/src/main/java/org/apache/accumulo/master/TimeoutTaskExecutor.java

```diff
@@ -0,0 +1,235 @@
+   */
+  public TimeoutTaskExecutor(int numThreads, long timeout, int expectedNumCallables) {
+    Preconditions.checkArgument(numThreads >= 1, "Number of threads must be at least 1.");
+    Preconditions.checkArgument(expectedNumCallables >= 0, "The expected number of callables must be non-negative.");
+
+    this.executorService = Executors.newFixedThreadPool(numThreads);
+    this.startedTasks = new ArrayBlockingQueue<>(expectedNumCallables);
+    this.timeout = timeout;
+    this.wrappedTasks = new ArrayList<>(expectedNumCallables);
+  }
+
+  /**
+   * Submits a new task to the executor.
+   *
+   * @param callable Task to run
+   */
+  public void submit(C callable) {
+    WrappedTask wt = new WrappedTask(callable);
+    wt.future = executorService.submit(wt);
+    wrappedTasks.add(wt);
+  }
+
+  /**
+   * Registers the callback to use on successful tasks.
+   *
+   * @param successCallback The callback function to invoke on success.
+   * @throws NullPointerException when a null successCallback is passed in
+   */
+  public void onSuccess(SuccessCallback<T,C> successCallback) {
+    this.successCallback = Objects.requireNonNull(successCallback, "Must provide a non-null successCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that throw exceptions.
+   *
+   * @param exceptionCallback The callback function to invoke on exceptions.
+   * @throws NullPointerException when a null exceptionCallback is passed in
+   */
+  public void onException(ExceptionCallback<C> exceptionCallback) {
+    this.exceptionCallback = Objects.requireNonNull(exceptionCallback, "Must provide a non-null exceptionCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that time out.
+   *
+   * @param timeoutCallback The callback function to invoke on timeouts.
+   * @throws NullPointerException when a null timeoutCallback is passed in
+   */
+  public void onTimeout(TimeoutCallback<C> timeoutCallback) {
+    this.timeoutCallback = Objects.requireNonNull(timeoutCallback, "Must provide a non-null timeoutCallback.");
+  }
+
+  /**
+   * Completes all the current tasks by dispatching to the appropriate callback.
+   *
+   * @throws IllegalStateException If all of the callbacks were not registered before calling this method.
+   * @throws InterruptedException  If interrupted while awaiting callable results.
+   */
+  public void complete() throws InterruptedException {
+    Preconditions.checkState(successCallback != null, "Must set a success callback before completing " + this);
+    Preconditions.checkState(exceptionCallback != null, "Must set an exception callback before completing " + this);
+    Preconditions.checkState(timeoutCallback != null, "Must set a timeout callback before completing " + this);
+
+    while (hasUnfinishedTasks()) {
+      completeTask(startedTasks.take());
+    }
+
+    wrappedTasks.clear();
+  }
+
+  private boolean hasUnfinishedTasks() {
+    for (WrappedTask wt : wrappedTasks) {
```

Comment:
`wrappedTasks` is not a sync list so if this method runs in a different thread than calls to `submit()` then changes may not be seen.  Could `wrappedTask` be replaced with an AtomicLong?  Increment the long before adding to executor and make `call()` decrement in try with finally block.  This method could check if the AtomicLong is greater than zero.

Reply:
> I like keeping it simpler as it solves the immediate use case for the Master status collections.

I agree with that.   I didn't realize only a single thread would access wrappedTask.  I am ok with leaving as is.  I was thinking the counter may be more efficient than repeatedly iterating over the list.  However the list is less bug prone.  I am ok with leaving as is.

Reply:
So I left it not thread safe but migrated to the counter. It'll reduce the complexity of the unfinished check by an order of magnitude. Definitely seems worth it.